### PR TITLE
Testing part 4

### PR DESCRIPTION
**Pull request by muffintoxin**
tests for form and dropzone components

### CODE REVIEW

File: test/dropzone/index.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+  it('deactivate', () => {
+    const wrapper = mount(createElement(Dropzone));
+    const instance = wrapper.instance();
+    instance.handleDragStart = fake();
+    instance.handleDragEnd = fake();
+    instance.handleGlobalDrop = fake();
+
+    const map = {};
+    document.removeEventListener = fake((event, cb) => {
```
[the remark this answers is not in the document]
Comment (muffintoxin) [reply test/dropzone/index.js]: `+`

File: test/dropzone/index.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+  it('activate', () => {
+    const wrapper = mount(createElement(Dropzone));
+    const instance = wrapper.instance();
+    instance.handleDragStart = fake();
+    instance.handleDragEnd = fake();
+    instance.handleGlobalDrop = fake();
+
+    const map = {};
+    document.addEventListener = fake((event, cb) => {
```
[the remark this answers is not in the document]
Comment (muffintoxin) [reply test/dropzone/index.js]: `+`

File: test/form/index.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    instance.addInput({ setPristine });
+    instance.setPristine(callback);
+
+    assert.deepEqual(instance.state, instance.getDefaultState(), 'state is back to default');
+    assert.isTrue(setPristine.calledOnce, 'setPristine on component was called');
+
+    setTimeout(() => {
+      assert.isTrue(callback.calledOnce, 'callback was called');
+    }, 100);
```
Comment (atomiomi) [inline test/form/index.js]: try `process.nextTick`

File: test/form/index.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    assert.lengthOf(error, 1, 'error was rendered');
+    assert.equal(error.text(), 'Error', 'formError text was rendered correct');
+
+    assert.lengthOf(button, 1, 'button was rendered');
+    assert.equal(button.text(), 'button', 'button text  was rendered correct');
+  });
+
+  it('setPristine', () => {
+    const wrapper = mount(createElement(Form));
```
Comment (atomiomi) [inline test/form/index.js]: try `shallow` and in other cases too

File: test/form/index.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    assert.deepEqual(instance.state, instance.getDefaultState(), 'state is back to default');
+    assert.isTrue(reset.calledOnce, 'reset on component was called');
+  });
+
+  it('submit', () => {
+    const preventDefault = fake();
+    const wrapper = shallow(createElement(Form));
+
+    wrapper.find('form').simulate('submit', { preventDefault });
+    const onSubmit = wrapper.props().onSubmit;
```
Comment (atomiomi) [inline test/form/index.js]: why test onSubmit like this?? check like other callback props - pass fake functions as props and check that they were called. check `onInvalidSubmit`, `onValidSubmit ` as well

File: test/form/index.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    assert.isTrue(preventDefault.calledOnce, 'preventDefault was called');
+    assert.exists(onSubmit, 'onSubmit prop exists');
+    assert.typeOf(onSubmit, 'function', 'onSubmit must be a funciton');
+  });
+
+  it('validate', () => {
+    const wrapper = shallow(createElement(Form));
+    const instance = wrapper.instance();
```
Comment (atomiomi) [inline test/form/index.js]: - create 2 instances
- add object with fake `validate` that returns resolved promise to 1 instance
- add another object with  fake `validate` that returns rejected promise to 2 instance
- test this 2 instances.

File: test/form/index.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    instance.addInput('input1');
+    instance.addInput('input2');
+    instance.addInput('input1');
+    instance.addInput('input2');
+    instance.addInput('input3');
+
+    assert.equal(instance.inputs.length, 3, 'only not existing input were added');
```
Comment (atomiomi) [inline test/form/index.js]: check array items - `['input1', 'input2', 'input3']`

File: test/form/index.js [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+    instance.addInput('input1');
+    instance.addInput('input2');
+    instance.addInput('input3');
+
+    instance.removeInput('input2');
+    instance.removeInput('input3');
+
+    assert.equal(instance.inputs.length, 1, 'input were removed');
+    assert.equal(instance.inputs[0], 'input1', 'correct input stayed');
```
Comment (atomiomi) [inline test/form/index.js]: just `deepEqual`

File: test/dropzone/index.js [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    assert.isFalse(props.onSelect.calledOnce, 'onSelect was not called');
+    assert.isTrue(props.onDrop.calledOnce, 'onDrop was called');
+
+    // with files
+    const wrapperWithoutFiles = shallow(createElement(Dropzone, props));
+    const instanceWithoutFiles = wrapperWithoutFiles.instance();
+
+    instanceWithoutFiles.handleDrop({ preventDefault, dataTransfer: { files: ['file1', 'file2'] } });
+    assert.isTrue(props.onSelect.calledOnce, 'onSelect was called');
+    assert.isFalse(props.onDrop.calledOnce, 'onDrop was not called');
```
[the remark this answers is not in the document]
Comment (muffintoxin) [reply test/dropzone/index.js]: `if (files.length) invoke(this.props.onSelect, files);
    invoke(this.props.onDrop, event);`

File: test/form/index.js (hunk 8 above)
Comment (muffintoxin) [reply test/form/index.js]: `+`

File: test/dropzone/index.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+    assert.equal(labelDrag.text(), 'labelDrag', 'drag label text was rendered');
+    assert.equal(labelRelease.text(), 'labelRelease', 'release label was rendered');
+
+    // default state
+    assert.isFalse(instance.state.isActive, 'default state for isActive is correct');
+    assert.isFalse(instance.state.isHover, 'default state for isHover is correct');
+  });
+
+  // componentDidMount
```
[the remark this answers is not in the document]
Comment (muffintoxin) [reply test/dropzone/index.js]: `+`

File: test/dropzone/index.js [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    assert.deepEqual(instance.state, instance.getDefaultState(), 'state is back to default');
+  });
+
+  it('activate', () => {
+    const wrapper = shallow(createElement(Dropzone));
+    const instance = wrapper.instance();
+    instance.handleDragStart = fake();
+    instance.handleDragEnd = fake();
+    instance.handleGlobalDrop = fake();
+    const originalListener = document.addEventListener;
```
[the remark this answers is not in the document]
Comment (muffintoxin) [reply test/dropzone/index.js]: `+`

File: test/dropzone/index.js [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+    const map = {};
```
[the remark this answers is not in the document]
Comment (muffintoxin) [reply test/dropzone/index.js]: `+`

File: test/form/index.js (hunk 3 above)
Comment (muffintoxin) [reply test/form/index.js]: `+`

File: test/form/index.js (hunk 4 above)
Comment (muffintoxin) [reply test/form/index.js]: `+`

File: test/form/index.js (hunk 7 above)
Comment (muffintoxin) [reply test/form/index.js]: `+`

File: test/form/index.js (hunk 5 above)
Comment (muffintoxin) [reply test/form/index.js]: `+`

File: test/form/index.js [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+    assert.isFalse(instance.state.isValid, 'state was changed');
+
+    instance.addInput({ setPristine });
+    instance.setPristine(callback);
+
+    assert.deepEqual(instance.state, instance.getDefaultState(), 'state is back to default');
+    assert.isTrue(setPristine.calledOnce, 'setPristine on component was called');
+
+    process.nextTick(() => {
+      assert.isTrue(callback.calledOnce, 'callback was called');
```
Comment (atomiomi) [inline test/form/index.js]: Test is async, use `done`

File: test/form/index.js [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+    const instance = wrapper.instance();
+    instance.validate = spy(() => Promise.resolve());
+
+    wrapper.find('form').simulate('submit', { preventDefault });
+
+    assert.isTrue(preventDefault.calledOnce, 'preventDefault was called');
+    assert.isTrue(instance.validate.calledOnce, 'validate was called');
+    assert.isTrue(props.onSubmit.calledOnce, 'onSubmit was called');
+
+    setTimeout(() => {
```
Comment (atomiomi) [inline test/form/index.js]: Use `process.nextTick`, and this test is async, use `done`

File: test/form/index.js [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+      assert.isTrue(props.onValidSubmit.calledOnce, 'onValidSubmit was called');
+      assert.isTrue(props.onInvalidSubmit.calledOnce, 'onInvalidSubmit was called');
```
Comment (atomiomi) [inline test/form/index.js]: How does one submit call both valid and invalid submit?

File: test/form/index.js [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+    instance.addInput('input1');
+    instance.addInput('input2');
+    instance.addInput('input3');
+
+    instance.removeInput('input2');
+    instance.removeInput('input3');
+
+    assert.equal(instance.inputs.length, 1, 'input were removed');
+    assert.deepEqual(instance.inputs[0], 'input1', 'correct input stayed');
```
Comment (atomiomi) [inline test/form/index.js]: i ment `deepEqual(instace.inputs, [‘input1’])`

File: test/form/index.js (hunk 14 above)
Comment (muffintoxin) [reply test/form/index.js]: keeped setTimeout

File: test/form/index.js [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    const instance = wrapper.instance();
+
+    instance.addInput('input1');
+    instance.addInput('input2');
+    instance.addInput('input3');
+
+    instance.removeInput('input2');
+    instance.removeInput('input3');
+
+    assert.equal(instance.inputs.length, 1, 'input were removed');
```
Comment (atomiomi) [inline test/form/index.js]: no need in this with deepEqual

**Pull request merged**